drivers/modem/modem_cellular.c

```diff
@@ -0,0 +1,1167 @@
+		return;
+	}
+
+	if (strlen(argv[1]) != 15) {
+		return;
+	}
+
+	for (uint8_t i = 0; i < 15; i++) {
+		data->imei[i] = argv[1][i] - '0';
+	}
+}
+
+static void modem_cellular_chat_on_cgmm(struct modem_chat *chat, char **argv, uint16_t argc,
+					void *user_data)
+{
+	struct modem_cellular_data *data = (struct modem_cellular_data *)user_data;
+
+	if (argc != 2) {
+		return;
+	}
+
+	strncpy(data->hwinfo, argv[1], sizeof(data->hwinfo) - 1);
+}
+
+static void modem_cellular_chat_on_creg(struct modem_chat *chat, char **argv, uint16_t argc,
+					void *user_data)
+{
+	struct modem_cellular_data *data = (struct modem_cellular_data *)user_data;
+
+	if (argc != 3) {
+		return;
+	}
+
+	data->access_tech = atoi(argv[1]);
+	data->registration_status = atoi(argv[2]);
+}
+
+static void modem_cellular_chat_on_cgatt(struct modem_chat *chat, char **argv, uint16_t argc,
+					 void *user_data)
+{
+	struct modem_cellular_data *data = (struct modem_cellular_data *)user_data;
+
+	if (argc != 2) {
+		return;
+	}
+
+	data->packet_service_attached = atoi(argv[1]);
+}
+
+MODEM_CHAT_MATCH_DEFINE(ok_match, "OK", "", NULL);
+MODEM_CHAT_MATCH_DEFINE(imei_match, "", "", modem_cellular_chat_on_imei);
+MODEM_CHAT_MATCH_DEFINE(cgmm_match, "", "", modem_cellular_chat_on_cgmm);
+MODEM_CHAT_MATCH_DEFINE(creg_match, "+CREG: ", ",", modem_cellular_chat_on_creg);
+MODEM_CHAT_MATCH_DEFINE(cgatt_match, "+CGATT: ", ",", modem_cellular_chat_on_cgatt);
+
+MODEM_CHAT_MATCHES_DEFINE(abort_matches, MODEM_CHAT_MATCH("ERROR", "", NULL));
+
+MODEM_CHAT_MATCHES_DEFINE(connect_abort_matches,
+			  MODEM_CHAT_MATCH("ERROR", "", NULL),
+			  MODEM_CHAT_MATCH("BUSY", "", NULL),
+			  MODEM_CHAT_MATCH("NO ANSWER", "", NULL),
+			  MODEM_CHAT_MATCH("NO CARRIER", "", NULL),
+			  MODEM_CHAT_MATCH("NO DIALTONE", "", NULL));
+
+MODEM_CHAT_SCRIPT_CMDS_DEFINE(init_chat_script_cmds,
+			      MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT", 100),
+			      MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT", 100),
+			      MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT", 100),
+			      MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT", 100),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("ATE0", ok_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CMEE=1", ok_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CREG=0", ok_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CGSN", imei_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("", ok_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CGMM", cgmm_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("", ok_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT+CMUX=0,0,5,127,10,3,30,10,2",
+							      100));
+
+MODEM_CHAT_SCRIPT_DEFINE(init_chat_script, init_chat_script_cmds, abort_matches,
+			 modem_cellular_chat_callback_handler, 10);
+
+MODEM_CHAT_SCRIPT_CMDS_DEFINE(net_stat_chat_script_cmds,
+			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CREG?", creg_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("", ok_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CGATT?", cgatt_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("", ok_match));
```

Comment:
This may be asking a lot but I think this base implementation should be changed to not poll this. Lets set CREG up for unsolicited events and react on the event. This would be a much better implementation and promote users making new drivers to not stick with polling methods. 
I think it's also a better example because it shows how to deal with unsolicited events in the modem chat.

Reply:
Deal